webauthn-authenticator-rs/src/MacAuthn/Sources/MacAuthn/MacAuthn.swift

```diff
@@ -0,0 +1,194 @@
+    case let .ok(data):
+        return String(data: try! JSONSerialization.data(withJSONObject: ["data": data]), encoding: .utf8)!
+    case let .error(message):
+        return String(data: try! JSONSerialization.data(withJSONObject: ["error": message]), encoding: .utf8)!
+    }
+}
+
+@_cdecl("perform_register")
+public func performRegister(options: SRString) -> SRString {
+    let options = try! JSONDecoder().decode(PublicKeyCredentialCreationOptions.self, from: Data(options.toArray()))
+    
+    let platformProvider = ASAuthorizationPlatformPublicKeyCredentialProvider(relyingPartyIdentifier: options.rp.id)
+    let platformKeyRequest = platformProvider.createCredentialRegistrationRequest(challenge: options.challenge.decodeBase64Url()!, name: options.user.name, userID: options.user.id.decodeBase64Url()!)
+    platformKeyRequest.displayName = options.user.displayName
+    platformKeyRequest.userVerificationPreference = ASAuthorizationPublicKeyCredentialUserVerificationPreference.init(rawValue: options.authenticatorSelection.userVerification ?? "preferred")
+    
+    let securityKeyProvider = ASAuthorizationSecurityKeyPublicKeyCredentialProvider(relyingPartyIdentifier: options.rp.id)
+    
+    let securityKeyRequest = securityKeyProvider.createCredentialRegistrationRequest(challenge: options.challenge.decodeBase64Url()!, displayName: options.user.displayName, name: options.user.name, userID: options.user.id.decodeBase64Url()!)
+    
+    securityKeyRequest.credentialParameters = []
+    for publicKeyParam in options.pubKeyCredParams {
+        let algorithm = ASCOSEAlgorithmIdentifier(rawValue: publicKeyParam.alg)
+        let parameters = ASAuthorizationPublicKeyCredentialParameters(algorithm: algorithm)
+        securityKeyRequest.credentialParameters.append(parameters)
+    }
+    
+    securityKeyRequest.excludedCredentials = []
+    for credential in (options.excludeCredentials ?? []) {
+        let id = credential.id.decodeBase64Url()!
+        let transports = credential.transports?.map {
+            ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport.init(rawValue: $0)
+        } ?? ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport.allSupported
+        let credential = ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor(credentialID: id, transports: transports)
+        securityKeyRequest.excludedCredentials.append(credential)
+    }
+    
+    securityKeyRequest.attestationPreference = ASAuthorizationPublicKeyCredentialAttestationKind(rawValue: options.attestation ?? "none")
+    securityKeyRequest.userVerificationPreference = ASAuthorizationPublicKeyCredentialUserVerificationPreference.init(rawValue: options.authenticatorSelection.userVerification ?? "preferred")
+    
+    if options.authenticatorSelection.requireResidentKey == true {
+        securityKeyRequest.residentKeyPreference = .required
+    } else {
+        securityKeyRequest.residentKeyPreference = .preferred
```

Comment:
This should use the value of `AuthenticatorSelectionCriteria.resident_key`; and default to `discouraged` if `require_resident_key = false`: https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-residentkey

At the moment, this will *always* try to use resident keys if a key supports them, potentially filling the authenticator's limited storage with useless RKs.

Reply:
Agreed, this needs to default to discouraged. Because the current code here effectively is forcing "true" on all code paths.